File: day1/day1.ts

if (Deno.args.length < 1) {
    console.log("Usage: deno "+Deno.mainModule.substring(Deno.mainModule.lastIndexOf('/') + 1)+" <input>");
    Deno.exit(1);
}
const input = await Deno.readTextFile(Deno.args[0]);


const lines = input.split("\n");
const left = [];
const right = [];
const instances:number[] = [];

for (const line of lines) {
    const w = line.split(/\s+/g);
    if (w.length > 1) {
        //console.log(w)
        const [a, b] = w;
        left.push(parseInt(a));
        right.push(parseInt(b));
        instances[parseInt(b)] = (instances[parseInt(b)] ?? 0) + 1;
    }
}

left.sort();
right.sort();

const distances:number[] = [];
const similars:number[] = [];
let totalDistance = 0;
let totalSimilar = 0;
for (let i = 0; i < left.length; i++) {
    const distance = Math.abs(right[i] - left[i]);
    distances.push(distance);
    totalDistance += distance;

    const sim = left[i] * (instances[left[i]] ?? 0);
    similars.push(sim);
    totalSimilar += sim;
}

//console.log(left, right, distances, instances, similars);
console.log("Total Distance:", totalDistance);
console.log("Total Similar:", totalSimilar);

File: day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path"
	"slices"
	"strconv"
	"strings"
)

func IAbs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage:", path.Base(os.Args[0]), "<input file>")
		os.Exit(1)
	}
	inFile, err := os.Open(os.Args[1]) //("day1-sample.txt")
	if err != nil {
		//log.Fatal(err)
		fmt.Println("Error: ", err)
		os.Exit(1)
	}
	defer inFile.Close()

	left := []int{}
	right := []int{}

	instances := make(map[int]int)

	// NOTE: Scanner has a 64k limit by default
	scanner := bufio.NewScanner(inFile)
	for scanner.Scan() {
		line := scanner.Text()
		//tokens := strings.Split(line, "   ")	// hacky solution which is correct (3 spaces), but I'd rather tokenize by any whitespace
		tokens := strings.Fields(line)  // I like this better, it tokenizes by any whitespace
		a, _ := strconv.Atoi(tokens[0]) // Boo, I wish this didn't need to be broken up over 2 lines
		left = append(left, a)
		b, _ := strconv.Atoi(tokens[1])
		right = append(right, b)
		//fmt.Println(tokens, len(tokens))
		instances[b] = instances[b] + 1 // In go, items that are not found in a map are the default value of the type (i.e. 0 for int)
	}
	//fmt.Println(left, right)
	slices.Sort(left)
	slices.Sort(right)
	//fmt.Println(left, right)

	totalDistance := 0
	totalSimilar := 0

	for i := 0; i < len(left); i++ {
		totalDistance += IAbs(right[i] - left[i])

		sim := left[i] * instances[left[i]]
		totalSimilar += sim
	}

	fmt.Println("Total Distance: ", totalDistance)
	fmt.Println("Total Similar: ", totalSimilar)
}
